onnx/shape_inference/implementation.cc

```diff
@@ -425,67 +382,172 @@ static void InferShapesImpl(
     }
-    fail_shape_inference(full_errors);
   }
+
+  void process(GraphProto& graph) {
+    GraphInferenceContext graph_inference_context{
+        value_types_by_name, opset_imports, symbol_table, schema_registry, ir_version, model_local_functions_map};
+    for (auto& vi : *graph.mutable_value_info()) {
+      updateType(vi);
+    }
+    for (auto& vi : *graph.mutable_input()) {
+      updateType(vi);
+    }
+    for (auto& vi : *graph.mutable_output()) {
+      updateType(vi);
+    }
+    for (const auto& tp : graph.initializer()) {
+      TypeProto initializer_type;
+      TypeProto_Tensor* initializer_tensor_type = initializer_type.mutable_tensor_type();
+      initializer_tensor_type->set_elem_type(tp.data_type());
+      // set the shape according to the initializer shape info
+      auto* shape = initializer_tensor_type->mutable_shape();
+      for (int i = 0; i < tp.dims_size(); ++i) {
+        shape->add_dim()->set_dim_value(tp.dims(i));
+      }
+      processInitializer(tp.name(), tp, initializer_type, input_data_by_name);
+    }
+    for (const auto& tp : graph.sparse_initializer()) {
+      TypeProto initializer_type;
+      auto* initializer_sparse_tensor_type = initializer_type.mutable_sparse_tensor_type();
+      initializer_sparse_tensor_type->set_elem_type(tp.values().data_type());
+      // set the shape according to the initializer shape info
+      auto* shape = initializer_sparse_tensor_type->mutable_shape();
+      for (int i = 0; i < tp.dims_size(); ++i) {
+        shape->add_dim()->set_dim_value(tp.dims(i));
+      }
+      processInitializer(tp.values().name(), tp, initializer_type, input_sparse_data_by_name);
+    }
+    // Collect data from constant nodes and check if any experimental ops exist
+    for (auto& n : *graph.mutable_node()) {
+      preprocess(n); // process constant node
+    }
+    for (auto& n : *graph.mutable_node()) {
+      process(n);
+    }
+    // Throw shape inference error if any. Error mode right now only supports 0 and 1.
+    // When set to 0, any node level shape inference errors are not thrown. This is to support backward compatiblity
+    // with 1.7 and earlier releases. When set to 1 it will throw all exceptions.
+    // TODO: Add a more granular way for exception handling.
+    if (options.error_mode > 0 && !inference_errors.empty()) {
+      std::string full_errors = "Shape inference error(s): ";
+      for (const std::string& error : inference_errors) {
+        full_errors += error + "\n";
+      }
+      fail_shape_inference(full_errors);
+    }
+  }
+
+ public:
+  ShapeInferenceImplBase(
+      GraphProto* gp,
+      const std::unordered_map<std::string, TypeProto*>& outer_scope_value_types_by_name_in,
+      const std::unordered_map<std::string, int>& opset_imports_in,
+      const ShapeInferenceOptions& options_in,
+      SymbolTable* symbol_table_in,
+      const ModelLocalFunctionsMap& model_local_functions_map_in,
```

Comment:
It would be great if we can add generated_shape_data_by_name_in into ShapeInferenceImplBase like [here](https://github.com/onnx/onnx/blob/67f21f2095084f96ca1698ed621a125fe8b53fb0/onnx/shape_inference/implementation.cc#L572) for FunctionProto.

Reply:
After discussion, we think it's fine to do it in another PR.